Zero-shot Computer Vision - Introduction (#1)
mmhamdy/computer-vision-course: chapters/en/Unit 11  - Zero Shot Computer Vision/1.mdx

> # Introduction

Reviewer: Title does not match the content for this paragraph. Maybe something like "Note" or "Before we start"?

Author: This will be the title of the chapter, not this section. remember our discussion about chapter organization and comparison to HF course?

Author: This will be the title of the chapter, not this section. remember our discussion about chapter organization and comparison to HF course?

Besides fundamentals, this unit assumes familiarity with concepts in transfer and multimodal learning. If you are not viewing the course sequentially in the provided order, we suggest at least reading the transfer learning section in unit 2, and the whole of unit 4.

> ## On Generalization

Reviewer: I suggest a more descriptive title -- The Concept of Generalization

Author: This is going to be too promising, and I just mention generalization in brief.

> We have now trained our model like a student cramming all epochs for a test. Now, the real test begins! We hope this knowledge translates beyond the specific pictures it learned from, allowing it to recognize unseen cats like Alice's and Ted’s furry friends. Think of it as the model learning the essence of catness, not just those specific furry faces it saw during training. This ability to apply knowledge to new situations is called generalization, and it's what separates a good model from a mere cat picture memorizer. Can you imagine an alternate universe without generalization? Yes, it’s pretty simple actually, you’ll only have to train your model on ALL the images of cats in the world (assuming they only exist on earth), including Alice’s and Ted’s, then find a way to prevent the current cat population from breeding. So, yeah, no big deal.

Reviewer: ```suggestion
We can train a standard cat classification model like a student cramming all epochs for a test. Now, the real test begins! We hope this knowledge translates beyond the specific pictures it learned from, allowing it to recognize unseen cats like Alice's and Ted’s furry friends. Think of it as the model learning the essence of catness, not just those specific furry faces the model saw during training. This ability to apply knowledge to new situations is called generalization, and it's what separates a good cat model from a mere cat picture memorizer. Can you imagine an alternate universe without generalization? Yes, it’s pretty simple actually, you’ll only have to train your model on ALL the images of cats in the world (assuming they only exist on earth), including Alice’s and Ted’s, then find a way to prevent the current cat population from breeding. So, yeah, no big deal.
```

Author: The first sentence serves as a continuation of other units, where lots of time has been spent on talking about training models. But I think referring to it as a cat model won't hurt.

So, we’ve gone from generalization, recognizing the unseen Alice’s and Ted’s cat pictures, to domain adaptation, recognizing animated cat pictures. But we’re much Greedier than that. You don’t want your model to be able to only recognize your cat pictures, or Alice’s and Ted’s, or not even cartoon cat! Having a model trained on cat pictures, you also want it to recognize pictures of llamas and falcons.

Well, now we’re in the turf of Zero-shot Learning. Also known as, ZSL.

[comment]: # (TODO: Illustration showing the difference between generalization, domain adaptation, and ZSL)

## What is Zero-shot Learning?

Let’s warm up with a definition

Zero-shot learning is a setup in which the model is presented with images belonging to **only** classes that it was not exposed to during training, at test time. In other words, the training and testing sets are **disjoint**.

Just a heads-up: in the classic ZSL setup, the test set only has pictures of classes the model hasn't seen before, not a single one from its training days. This may seem a little bit unrealistic, it's  like asking a student to ace an exam on only materials they've never studied. 

> Luckily, there's a more pragmatic version of ZSL that doesn't have this strict rule and is called generalized zero-shot learning, or GZSL. This more flexible approach allows the test set to include both seen and unseen classes. It's a more realistic scenario, reflecting how things work in the real world.

Reviewer: I think this section did not fully explain the reason to distinguish between ZSL (with only unseen test cases) vs GZSL (with seen + unseen test cases).

Author: That's because there is a whole section for it in your chapter 😉

Let’s have a nice outline for the most important moments in ZSL below

| 2008 | The first sparks of (the question of) zero-shot learning                |
| 2009 | The term zero-shot learning was coined                                  |
| 2013 | The concept of Generalized ZSL was introduced[^2]                       |
| 2017 | The first application of the encoder-decoder paradigm to ZSL[^3]        |
| 2020 | OpenAI’s GPT-3 achieves an impressive performance on zero-shot NLP[^4]  |
| 2021 | OpenAI’s CLIP takes zero-shot computer vision to a whole new level[^5]  |

> The impact of CLIP has been profound, ushering in a new era of zero-shot computer vision research. Its versatility and performance have opened up exciting new possibilities. It could be said that the history of zero-shot computer vision can be divided into two eras: The **pre-CLIP** and **post-CLIP** eras.

Reviewer: Maybe at the end of section we can add a few sentences to point out there is a chapter for CLIP .

Author: I don't think we're going to have a whole chapter for CLIP, For two reasons. First, CLIP will be explained in much detail in other units. Second, we don't have the capacity to finish already existing chapters. So, we are going to distribute CLIP among the coming chapters and mention it in context.

Let’s begin by answering a simple question, *how can we humans do it? How can we recognize objects that we have not seen before?*

Yes, you’re totally right! We need some **other information** about that object. We can easily identify a tiger, even if we haven’t seen one before, given that we have a knowledge of what a lion looks like. A tiger is a lion with stripes and minus the mane. A zebra is a black and white striped version of a horse. Deadpool is a red-black version of Spiderman.

As we can see, zero-shot computer vision is essentially multi-modal. If generalization is like learning a language by reading books and talking to people. Zero-shot computer vision is like learning a language by reading a dictionary and listening to someone describe what it sounds like.

Can we think of zero-shot learning as an unsupervised learning problem? No, this is still supervised learning. In unsupervised learning, the model learns from **unlabeled data**. In zero-shot learning, the model learns from **dataless labels**, remember it was called **dataless classification** in its infancy.

> ### What Other Information?

Reviewer: I think we can give a more descriptive title for this section.

Author: This is too short to be its own section, I guess. I don't have additional details to add to this part.

3. **Class label vectors**: Simply using the embeddings of class labels.

[comment]: # (TODO: Illustration showing the different forms of semantic information)

And using this semantic information, you train a model to learn a mapping function between the image features and the semantic features. At inference time, the model predicts the class label by looking for the closest label in the semantic space by using, for example, k-nearest neighbor. We can say that we are transferring knowledge obtained from seen classes with the help of semantic information.

Different zero-shot computer vision methods differ in what **semantic information** they use and which **embedding space** they utilize at inference.

[comment]: # (TODO: Illustration showing the pipeline of zero-shot learning)

### How Is This Different from Transfer Learning?

Good question! Zero-shot learning (ZSL) falls within the broader category of transfer learning, specifically under the umbrella of **heterogeneous transfer learning**. This means that ZSL relies on transferring knowledge gained from a **source domain** (seen classes) to a *distinct* **target domain** (unseen classes). 

> Due to the disparity between the source and target domains, ZSL faces specific challenges in transfer learning. These include overcoming domain shift, where the data distributions differ significantly, and handling the lack of labeled data in the target domain. But we’re going to talk about these challenges in detail later, so don’t worry about it now. For now, let’s talk a little about the different methods of zero-shot computer vision.

Reviewer: ```suggestion
Due to the disparity between the source and target domains, ZSL faces specific challenges in transfer learning. These include overcoming domain shift, where the domain data distributions are distinct from one another, and handling the lack of labeled data in the target domain. But we’re going to talk about these challenges in detail later, so don’t worry about it now. For now, let’s talk a little about the different methods of zero-shot computer vision.
```

Author: Domain and data distribution are the same thing here. And I wanted to show that the difference is a significant one.

- **Embedding-based methods**: The model learns a common embedding space where both images and their semantic features/representations are projected. New unseen classes are then predicted using a similarity measure in that space. For example, CLIP.
- **Generative-based methods**: These methods utilize generative models to create synthetic images or visual features for the unseen classes based on the samples of seen classes and semantic representations of both classes. This allows for training models on these unseen classes even without real data. This way, we’ve cheated a little bit and converted the zero-shot problem into a supervised learning problem. For example, CVAE[^6].

[comment]: # (TODO: Diagram showing the categories of ZSL methods and the sub-categories)

The choice between embedding-based and generative-based approaches, like every other choice that you have to make in machine learning, *depends* on the specifics of the task at hand and the available resources. Embedding-based methods are often preferred for their efficiency and scalability, while generative-based methods offer greater flexibility and potential for handling complex data.

> But Anyway, In this unit we are going to address *only* embedding-based methods, and to be more specific **attention-based embedding-based methods**. Zero-shot learning is a big topic and a comprehensive coverage may require a course of its own. Interested readers may check the provided additional readings and have fun.

Reviewer: ```suggestion
But Anyway, In this unit we are going to address *only* embedding-based methods, including **attention-based embedding-based methods**. Zero-shot learning is a big topic and a comprehensive coverage may require a course of its own. Interested readers may check the provided additional readings and have fun.
```

Author: Yeah, much better and broader, just in case.

## Attack Of The Clones: How ZSL Is Different From X?

We’ve gone a long way! We now have a decent idea of zero-shot learning, its history, its application in computer vision, and how it generally works. To round out this introduction, we are going to compare ZSL to other different methods that may initially appear similar.

> - **Domain Adaptation (DA)**: This should be familiar by now. We can think of Zero-shot Learning as an extreme case of Domain Adaptation (greedy, greedy), dealing with the problem of learning to recognize unseen classes with no data at all. Domain Adaptation focuses on bridging the gap between two related domains (datasets) with different distributions and it requires labeled data.

Reviewer: ```suggestion
- **Domain Adaptation (DA)**: This should be familiar by now. We can think of Zero-shot Learning as an extreme case of Domain Adaptation, dealing with the problem of learning to recognize unseen classes with no data at all. Domain Adaptation focuses on bridging the gap between two related domains (datasets) with different distributions and it requires labeled data.
```

Reviewer: The greedy comment was a bit confusing for me. But I could be the only one who did not get it.  :)

Author: It's just a joke😅. A reference to greedy part mentioned in the first section.
